integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/highlight/HighlightProjectionBaseIT.java

```diff
@@ -0,0 +1,1765 @@
+	private static final String LEVEL2_SINGLE_EMPTY_OBJECT_DOCUMENT_ID = "level2SingleEmptyObjectDocId";
+	private static final String LEVEL2_NO_OBJECT_DOCUMENT_ID = "level2NoObjectDocId";
+
+	private static final String MISSING_LEVEL1_DOCUMENT_ID = "missingLevel1DocId";
+	private static final String MISSING_LEVEL1_SINGLE_VALUED_FIELD_DOCUMENT_ID = "missingLevel1SVFieldDocId";
+	private static final String MISSING_LEVEL2_DOCUMENT_ID = "missingLevel2DocId";
+	private static final String MISSING_LEVEL2_SINGLE_VALUED_FIELD_DOCUMENT_ID = "missingLevel2SVFieldDocId";
+
+	protected final DataSet dataSet;
+	private final RecursiveComparisonConfiguration recursiveComparisonConfig;
+
+	@ClassRule
+	public static final SearchSetupHelper setupHelper = new SearchSetupHelper();
+
+
+	public HighlightProjectionBaseIT(DataSet dataSet) {
+		this.dataSet = dataSet;
+		this.recursiveComparisonConfig = RecursiveComparisonConfiguration.builder().
+				withComparatorForType( Comparator.nullsFirst( Comparator.naturalOrder() ), BigDecimal.class )
+				.build();
+	}
+
+	@BeforeClass
+	public static void setup() {
+		setupHelper.start()
+				.withIndexes( mainIndex, missingLevel1Index,
+						missingLevel1SingleValuedFieldIndex,
+						missingLevel2Index,
+						missingLevel2SingleValuedFieldIndex
+				)
+				.setup();
+
+		BulkIndexer compositeForEachMainIndexer = mainIndex.bulkIndexer();
+		BulkIndexer compositeForEachMissingLevel1Indexer = missingLevel1Index.bulkIndexer();
+		BulkIndexer compositeForEachMissingLevel1SingleValuedFieldIndexer = missingLevel1SingleValuedFieldIndex.bulkIndexer();
+		BulkIndexer compositeForEachMissingLevel2Indexer = missingLevel2Index.bulkIndexer();
+		BulkIndexer compositeForEachMissingLevel2SingleValuedFieldIndexer = missingLevel2SingleValuedFieldIndex.bulkIndexer();
+		dataSets.forEach(
+				d -> d.contribute( mainIndex, compositeForEachMainIndexer,
+						missingLevel1Index, compositeForEachMissingLevel1Indexer,
+						missingLevel1SingleValuedFieldIndex,
+						compositeForEachMissingLevel1SingleValuedFieldIndexer,
+						missingLevel2Index, compositeForEachMissingLevel2Indexer,
+						missingLevel2SingleValuedFieldIndex,
+						compositeForEachMissingLevel2SingleValuedFieldIndexer
+				) );
+
+		compositeForEachMainIndexer.join( compositeForEachMissingLevel1Indexer,
+				compositeForEachMissingLevel1SingleValuedFieldIndexer, compositeForEachMissingLevel2Indexer,
+				compositeForEachMissingLevel2SingleValuedFieldIndexer
+		);
+	}
+
+	protected ProjectionFinalStep<List<String>> singleValuedHighlight(SearchProjectionFactory<?, ?> f,
+			String absoluteFieldPath) {
+		return f.highlight( absoluteFieldPath );
+	}
+
+	protected ProjectionFinalStep<List<String>> multiValuedHighlight(SearchProjectionFactory<?, ?> f,
+			String absoluteFieldPath) {
+		return f.highlight( absoluteFieldPath );
+	}
+
+	@Test
+	public void objectOnLevel1AndObjectOnLevel2() {
```

Comment:
yeah, can't say that I like it.. I did want to use the `AbstractProjectionInObjectProjectionIT` even though it would either mean having the test live in the projection package or changing the visibility of some constructors/fields. 

Now because the tests in `AbstractProjectionInObjectProjectionIT` are using `match_all`, which for highlighting means - "nothing to highlight", in the end, it leads to different results than what is expected from simple projection. 
Also, I'm not sure if you've seen that already, but from what it looks in the case of ES - nested results are "unusable" (meaning  we cannot reconstruct a nice nested structure from what we get): 

When I'd use a wildcard query, so every string gets highlighted, we end up with the:
```json
{
  "took": 10,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1.0,
    "hits": [
      {
        "_index": "missinglevel1field1-000001",
        "_id": "missingLevel1SVFieldDocId_analyzedString/single=FLATTENED/multi=FLATTENED",
        "_score": 1.0,
        "_routing": "analyzedString/single=FLATTENED/multi=FLATTENED",
        "_source": {
          "level1_flattened": [
            {
              "multiValuedField_analyzedString": [
                "string_0",
                "string_1"
              ],
              "level2": [
                {
                  "singleValuedField_analyzedString": "string_2",
                  "multiValuedField_analyzedString": [
                    "string_3",
                    "string_4"
                  ]
                },
                {
                  "singleValuedField_analyzedString": "string_5",
                  "multiValuedField_analyzedString": [
                    "string_6",
                    "string_7"
                  ]
                }
              ]
            },
            {
              "multiValuedField_analyzedString": [
                "string_8",
                "string_9"
              ],
              "level2": [
                {
                  "singleValuedField_analyzedString": "string_10",
                  "multiValuedField_analyzedString": [
                    "string_11",
                    "string_12"
                  ]
                },
                {
                  "singleValuedField_analyzedString": "string_13",
                  "multiValuedField_analyzedString": [
                    "string_14",
                    "string_15"
                  ]
                }
              ]
            }
          ],
          "_entity_type": "missingLevel1Field1Type"
        },
        "highlight": {
          "level1_flattened.multiValuedField_analyzedString": [
            "string_0"
          ],
          "level1_flattened.level2.singleValuedField_analyzedString": [
            "<em>string_2</em>",
            "<em>string_5</em>",
            "<em>string_10</em>",
            "<em>string_13</em>"
          ],
          "level1_flattened.level2.multiValuedField_analyzedString": [
            "string_3"
          ]
        }
      }
    ]
  }
}
```
but if we do the no-match: then the result is:
```json
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1.0,
    "hits": [
      {
        "_index": "missinglevel1field1-000001",
        "_id": "missingLevel1SVFieldDocId_analyzedString/single=FLATTENED/multi=FLATTENED",
        "_score": 1.0,
        "_routing": "analyzedString/single=FLATTENED/multi=FLATTENED",
        "highlight": {
          "level1_flattened.multiValuedField_analyzedString": [
            "string_0"
          ],
          "level1_flattened.level2.singleValuedField_analyzedString": [
            "string_2"
          ],
          "level1_flattened.level2.multiValuedField_analyzedString": [
            "string_3"
          ]
        }
      }
    ]
  }
}
```
so yeah ... I didn't see a simple way to match the results, and I wasn't sure that refactoring that `AbstractProjectionInObjectProjectionIT` to introduce even more methods to override (query, match results) by implementors was a good idea since that test is complex already. 🤷🏻‍♂️

Reply:
I'll try to answer your main points.

1. TBH I wouldn't be shocked if those tests moved to the projection package, since they're testing... a projection.
2. Good point about `match_all`. It's fine if you make that overridable in the base class and use a different query in highlighting tests. I think that wildcard query you mentioned gives satisfying results, or did I miss something?
3. If nesting cannot be supported on Elasticsearch, I would just disable those tests for Elasticsearch, and create a simpler, separate test to check the Elasticsearch-specific behavior.


By the way, if Elasticsearch cannot apply nesting to highlights, **we should probably forbid using highlighting in a nested context with Elasticsearch** (i.e. throw an exception if `f.highlight` is used within a `f.object` projection). That would at least avoid some confusion from users. We can check whether a `highlight` projection is nested in the `request` method of a projection: `context.absoluteCurrentFieldPath() == null` means you're in the root and everything is fine, otherwise you'd throw an exception. And now that I think about it... we should probably do this for all projections that are not affected by the nesting context, just to be safe. We can relax this rule later. I created https://hibernate.atlassian.net/browse/HSEARCH-4827 to address that; feel free to do whatever is easier for `highlight` projections in this PR.